segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -293,17 +295,54 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+                                            CompletableFuture<Void> retFuture = CompletableFuture.completedFuture(null);
+
+                                            // Add to list of chunks to delete
+                                            chunksToDelete.add(chunkMetadata.getName());
+
+                                            // Skip if metadata is possibly deleted in last attempt, we are done.
+                                            if (null == chunkMetadata) {
+                                                currentChunkName.set(null);
+                                                return retFuture;
+                                            }
+
+                                            // Add to batch and commit batch if required.
+                                            currentBatch.add(chunkMetadata);
+                                            if (chunkMetadata.isActive()) {
+                                                if (currentBatch.size() > config.getGarbageCollectionTransactionBatchSize()) {
+                                                    // Commit batch
+                                                    retFuture = addTransactionForUpdateBatch(currentBatch, streamSegmentName);
+                                                    // Clear batch
+                                                    currentBatch.clear();
+                                                }
+                                            }
+                                            // Move next
+                                            currentChunkName.set(chunkMetadata.getNextChunk());
+                                            return retFuture;
+                                        }, storageExecutor),
+                                storageExecutor)
+                                .thenComposeAsync( v -> {
+                                    if (currentBatch.size() > 0) {
+                                        return addTransactionForUpdateBatch(currentBatch, streamSegmentName);
+                                    }
+                                    return CompletableFuture.completedFuture(null);
+                                }, storageExecutor)
                                 .thenComposeAsync(v -> this.addChunksToGarbage(txn.getVersion(), chunksToDelete), storageExecutor)
+                                .thenComposeAsync(v -> deleteBlockIndexEntriesForSegment(streamSegmentName, segmentMetadata.getStartOffset(), segmentMetadata.getLength()))
                                 .thenComposeAsync(v -> {
-                                    deleteBlockIndexEntriesForChunk(txn, streamSegmentName, segmentMetadata.getStartOffset(), segmentMetadata.getLength());
-                                    txn.delete(segmentMetadata.getName());
-                                    return txn.commit();
+                                    @Cleanup
```

Comment:
This `@Cleanup` will close your `innerTxn` before it has a chance to execute everything it needs. It will call `close` after line 345 executes, however line 345 returns a CompletableFuture, which indicates there is an async task which will complete sometime later. How is that task execution going to be affected if `innerTxn` is in a `closed` state?

Reply:
Oops forgot to fix that. Good catch.
Fixed now.